### DISCUSSION TITLE
Nominatim Error 500 - Could not find Driver

### DISCUSSION THREAD

**Discussion by VBviniciusVB** [osm-search/Nominatim#3026] [Q&A] [answered]
Hello, this is my first comment on GitHub :)

**--- Description ---**
I'm installing Nominatim on a new server (Ubuntu 22.02), and I'm encountering problems when opening it to perform reverse geocoding, an error appears which I believe is related to drivers or Postgres 15, I've looked on GitHub and on sites but I didn't find something specific to help me.

**--- Problem ---**
When I try to open localhost, the following error screen appears:
![image](https://user-images.githubusercontent.com/17010127/229911485-47215c46-00b9-4444-aeb4-f2d56bc1598b.png)
Image description:
error
code 500
message "Failed to establish database connection:could not find driver

I made sure to go to the Nominatim website and update all the dependencies mentioned and it still doesn't seem to have changed anything.
https://nominatim.org/release-docs/latest/appendix/Install-on-Ubuntu-22/

**--- Additional Information ---**
Distributor ID: Ubuntu
Description: Ubuntu 22.04.2 LTS
Release: 22.04
Codename: jammy

postgres (PostgreSQL) 15.2 (Ubuntu 15.2-1.pgdg22.04+1)

PHP 8.2.4 (cli) (built: Mar 16 2023 14:34:48) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.2.4, Copyright (c) Zend Technologies
    with Zend OPcache v8.2.4, Copyright (c), by Zend Technologies

**--- Question ---**
How can I fix this problem where the error 500 appears? Would the problem be caused by my using the latest versions of Postgres/PHP?

**Comment by mtmail** [comment]
As far as I know several users use postgresql 15 and PHP 8. Not sure which exact version.

Which version of nominatim are you using? `nominatim --version`

Do you see the same error on command-line? `nominatim reverse --lat -23.5358 --lon -46.6532` and `nominatim status --format json`

**Reply by VBviniciusVB** [reply]
Hello!

I missed the info about nominatim sry..

> Nominatim version 4.2.2-0 ()


The command `nominatim reverse --lat -23.5358 --lon -46.6532` works fine

> 2023-04-04 17:42:56: Using project directory: /srv/nominatim/Nominatim-4.2.2
> {"place_id":5289170,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"node","osm_id":8503460090,"lat":"-23.5359986","lon":"-46.6530639","place_rank":30,"category":"highway","type":"bus_stop","importance":9.99999999995449e-6,"addresstype":"highway","name":"Rua Das Palmeiras, 467","display_name":"Rua Das Palmeiras, 467, Rua das Palmeiras, Santa Cecília, São Paulo, Região Imediata de São Paulo, Região Metropolitana de São Paulo, Região Geográfica Intermediária de São Paulo, São Paulo, Região Sudeste, 01211-101, Brasil","address":{"highway":"Rua Das Palmeiras, 467","road":"Rua das Palmeiras","suburb":"Santa Cecília","city_district":"Santa Cecília","city":"São Paulo","municipality":"Região Imediata de São Paulo","county":"Região Metropolitana de São Paulo","state_district":"Região Geográfica Intermediária de São Paulo","state":"São Paulo","ISO3166-2-lvl4":"BR-SP","region":"Região Sudeste","postcode":"01211-101","country":"Brasil","country_code":"br"},"boundingbox":["-23.5360486","-23.5359486","-46.6531139","-46.6530139"]}

The command `nominatim status --format json` too

> 2023-04-04 17:44:47: Using project directory: /srv/nominatim/Nominatim-4.2.2
> {"status":0,"message":"OK","data_updated":"2023-03-29T20:15:13+00:00","software_version":"4.2.2-0","database_version":"4.2.2-0"}

**Reply by mtmail** [reply]
So something is different in the webserver environment. Which webserver do you use? https://nominatim.org/release-docs/latest/admin/Deployment/ mentions several. If you're lucky just restarting the server already works.

**Reply by VBviniciusVB** [reply]
Hello, the first attempt was made in **apache2** as shown above, and in **Nginx** it is just showing error 404 maybe it is just wrong configuration.. (Yes I downloaded both to test nominatim)
I tried and prayed to restart the system to make it work, but it didn't work either unfortunately..

The Nominatim directory is in `/srv/nominatim/website` I copied it to be in `/srv/nominatim/Nominatim-4.2.2/website` (Extracted file) but I don't know which one would be right, I tried putting apache2 for both configurations but both give the same problem, could you tell me which one is correct?

**Reply by mtmail** [reply]
The extracted files are most likely not the correct ones. Check that your `status.php` contains lines like `@define('CONST_` near the top. That's something Nominatim added during installation. You can also recreate the `website` directory by running `nominatim refresh --website`.

Try https://nominatim.org/release-docs/latest/admin/Deployment/#nominatim-with-apache again. You will need to replace "nominatim-project" with "nominatim". Restart the Apache webserver (`sudo service apache2 restart`). Check if there are any errors in `/var/log/apache2`.

**Reply by VBviniciusVB** [reply]
Hi!
inside the `status.php` file has all the @defines, here the code:

```
<?php

@define('CONST_Debug', $_GET['debug'] ?? false);
@define('CONST_LibDir', '/usr/local/lib/nominatim/lib-php');
@define('CONST_TokenizerDir', '/srv/nominatim/tokenizer');
@define('CONST_NominatimVersion', '4.2.2-0');

@define('CONST_Database_DSN', 'pgsql:dbname=nominatim');
@define('CONST_Default_Language', false);
@define('CONST_Log_DB', false);
@define('CONST_Log_File', false);
@define('CONST_NoAccessControl', true);
@define('CONST_Places_Max_ID_count', 50);
@define('CONST_PolygonOutput_MaximumTypes', 1);
@define('CONST_Search_BatchMode', false);
@define('CONST_Search_NameOnlySearchFrequencyThreshold', '500');
@define('CONST_Use_US_Tiger_Data', false);
@define('CONST_MapIcon_URL', false);

require_once('/usr/local/lib/nominatim/lib-php/website/status.php');
```

Here the nominatim commands (This is not cut in half) (I tested if this made it work but no results):

```
nominatim@evolusat:~$ nominatim refresh --website
2023-04-05 16:40:33: Using project directory: /srv/nominatim
2023-04-05 16:40:33: Setting up website directory at /srv/nominatim/website
nominatim@evolusat:~$ exit
sair
ubuntu@evolusat:~$ sudo -u www-data head -n 1 /srv/nominatim/website/search.php
<?php
ubuntu@evolusat:~$ 
```

I tested some other situations too, when I try to access `http://localhost/nominatim/status.php` I come across the following error: `ERROR: Database connection failed`, it seems that nominatim by apache2 cannot connect to Postgres ...

Would there be any relationship that must be made for Postgres to recognize apache2 or nominatim? The server I used previously was on Nominatim 3.6.0 and the installation process was different in some parts...

**Reply by mtmail** [reply]
Check if the user `www-data` can access the database. `sudo -u www-data psql nominatim`. If not then the database user might have to be created with `sudo -u postgres createuser www-data`

**Reply by VBviniciusVB** [reply]
Hello, running the commands show me it:

```
ubuntu@evolusat:~$ sudo -u www-data psql nominatim
[sudo] senha para ubuntu: 
could not change directory to "/home/ubuntu": Permissão negada
psql (15.2 (Ubuntu 15.2-1.pgdg22.04+1))
Type "help" for help.


nominatim=> exit
could not save history to file "/var/www/.psql_history": Arquivo ou diretório inexistente
ubuntu@evolusat:~$ sudo -u postgres createuser www-data
could not change directory to "/home/ubuntu": Permissão negada
createuser: error: creation of new role failed: ERROR:  role "www-data" already exists
ubuntu@evolusat:~$ sudo -i -u nominatim
nominatim@evolusat:~$ psql -U www-data -d nominatim
psql: error: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: FATAL:  Peer authentication failed for user "www-data"
nominatim@evolusat:~$ 
```

**Comment by lonvia** [comment]
The error message essentially means that PHP cannot find its postgres library/driver/module (php-pgsql). As @mtmail already suggested: make sure `nominatim admin --check-database` and `nominatim reverse --lat -23.5358 --lon -46.6532` work. Both commands should give you more useful output when they error out.

**Reply by VBviniciusVB** [reply]
Hello lonvia, I've already tried using these commands, both work without any problem (I believe), I will only use REVERSE GEOCODING, so I didn't give importance to wikipedia/wikidata like this in the Warning...

```
nominatim@evolusat:~$ nominatim admin --check-database
2023-04-06 09:17:45: Using project directory: /srv/nominatim
2023-04-06 09:17:45: Checking database
Checking database connection ... OK
Checking for placex table ... OK
Checking for placex content ... OK
Checking that tokenizer works ... OK
Checking for wikipedia/wikidata data ... WARNING

Wikipedia/Wikidata importance tables missing.
Quality of search results may be degraded. Reverse geocoding is unaffected.
See https://nominatim.org/release-docs/latest/admin/Import/#wikipediawikidata-rankings

Checking indexing status ... OK
Checking that database indexes are complete ... OK
Checking that all database indexes are valid ... OK
Checking TIGER external data table. ... not applicable
nominatim@evolusat:~$ nominatim reverse --lat -23.5358 --lon -46.6532
2023-04-06 09:17:58: Using project directory: /srv/nominatim
{"place_id":5289170,"licence":"Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright","osm_type":"node","osm_id":8503460090,"lat":"-23.5359986","lon":"-46.6530639","place_rank":30,"category":"highway","type":"bus_stop","importance":9.99999999995449e-6,"addresstype":"highway","name":"Rua Das Palmeiras, 467","display_name":"Rua Das Palmeiras, 467, Rua das Palmeiras, Santa Cecília, São Paulo, Região Imediata de São Paulo, Região Metropolitana de São Paulo, Região Geográfica Intermediária de São Paulo, São Paulo, Região Sudeste, 01211-101, Brasil","address":{"highway":"Rua Das Palmeiras, 467","road":"Rua das Palmeiras","suburb":"Santa Cecília","city_district":"Santa Cecília","city":"São Paulo","municipality":"Região Imediata de São Paulo","county":"Região Metropolitana de São Paulo","state_district":"Região Geográfica Intermediária de São Paulo","state":"São Paulo","ISO3166-2-lvl4":"BR-SP","region":"Região Sudeste","postcode":"01211-101","country":"Brasil","country_code":"br"},"boundingbox":["-23.5360486","-23.5359486","-46.6531139","-46.6530139"]}
nominatim@evolusat:~$ 
```

**Reply by lonvia** [reply]
Then your system setup is somewhere messed up. If I have to guess: `php-cli` is installed from Ubuntu and thus uses the standard PHP with its libraries from Ubuntu which you have installed. Nginx/Apache use whatever custom PHP you are running and for which you have not installed the libraries listed in [the manual](https://nominatim.org/release-docs/latest/admin/Installation/#prerequisites).

**Comment by VBviniciusVB** [comment, marked as answer]
Hello, greetings everyone!

I managed to solve the problem, I can't say exactly how I got to this point, but PHP in apache2 had the PHP 8.1 configuration enabled along with PHP 8.2, the system may have been lost and used the older version, resulting in the error demonstrating earlier... What I did was run the command:
```
sudo a2dismod
php8.1
```

Thanks for the support!
